### PR TITLE
Play nicer with ndarray subclasses

### PR DESCRIPTION
**Pull request by jcrist**
Two changes:

- Add `subok` kwarg to `da.from_array`.

Mirrors the `subok` kwarg in `np.array`. If True, array-like's will be passed through as chunks unchanged. Default is False. This allows avoiding the `np.asarray` call, making it easier to work with non numpy arrays as chunks.

To make this work we change `getarray` to a more generic `getter` function and add a `subok` parameter. Alternatively we could have kept the same format and created more getter functions, but this would have led to an explosion of functions for all the different combinations
(which seemed less maintainable).

- Add `da.asanyarray`

Similar to `da.asarray`, except avoids the call to `np.asarray`.

### CODE REVIEW

**Comment by jcrist** [comment]
I'm not sure if `subok` is the best keyword here, as technically we don't care if they're `np.ndarray` subclasses (e.g. works fine with `sparse.COO` which isn't). Perhaps `asarray` or something would be a better keyword?

**Comment by mrocklin** [comment]
> Perhaps asarray or something would be a better keyword?

No strong thoughts from me.  `subok` isn't immediately clear to me.

**Comment by jcrist** [comment]
I've switched to use `asarray=True` as the kwarg. Looks like github is having slowdowns processing events (new commit isn't showing up here), but you can see the commit on my branch. I'm happy to merge pending review/tests passing.

**Comment by jcrist** [comment]
Planning to merge later today if no comment.

File: dask/array/core.py [hunk 1]
```diff
@@ -45,41 +45,36 @@
...
-        return getarray(a, b2, lock)[b3]
+        return getter(a, b2, asarray=asarray, lock=lock)[b3]
 
     if lock:
         lock.acquire()
     try:
         c = a[b]
-        if type(c) != np.ndarray:
+        if asarray:
             c = np.asarray(c)
```
Comment (mrocklin) [inline dask/array/core.py]: Any value in making `asarray` a callable that defaults to `np.asarray`?

Comment (jcrist) [inline dask/array/core.py]: Working with it as a boolean makes the optimizations easier to write and reason about. Could be done with a callable, but it'd be a bit trickier. Perhaps hold off until a use case presents itself? I'm not sure when that would be useful.

**Comment by mrocklin** [comment]
Sure

**Pull request merged**